# Reading in the csv file and putting it in a dataframe

In [1]:
import pandas as pd

df = pd.read_csv('Tweets.csv')
df

tweet_id airline_sentiment  airline_sentiment_confidence  \
0      570306133677760513           neutral                        1.0000   
1      570301130888122368          positive                        0.3486   
2      570301083672813571           neutral                        0.6837   
3      570301031407624196          negative                        1.0000   
4      570300817074462722          negative                        1.0000   
...                   ...               ...                           ...   
14635  569587686496825344          positive                        0.3487   
14636  569587371693355008          negative                        1.0000   
14637  569587242672398336           neutral                        1.0000   
14638  569587188687634433          negative                        1.0000   
14639  569587140490866689           neutral                        0.6771   

               negativereason  negativereason_confidence         airline  \
0                         NaN                        NaN  Virgin America   
1                         NaN                     0.0000  Virgin America   
2                         NaN                        NaN  Virgin America   
3                  Bad Flight                     0.7033  Virgin America   
4                  Can't Tell                     1.0000  Virgin America   
...                       ...                        ...             ...   
14635                     NaN                     0.0000        American   
14636  Customer Service Issue                     1.0000        American   
14637                     NaN                        NaN        American   
14638  Customer Service Issue                     0.6659        American   
14639                     NaN                     0.0000        American   

      airline_sentiment_gold             name negativereason_gold  \
0                        NaN          cairdin                 NaN   
1                        NaN         jnardino                 NaN   
2                        NaN       yvonnalynn                 NaN   
3                        NaN         jnardino                 NaN   
4                        NaN         jnardino                 NaN   
...                      ...              ...                 ...   
14635                    NaN  KristenReenders                 NaN   
14636                    NaN         itsropes                 NaN   
14637                    NaN         sanyabun                 NaN   
14638                    NaN       SraJackson                 NaN   
14639                    NaN        daviddtwu                 NaN   

       retweet_count                                               text  \
0                  0                @VirginAmerica What @dhepburn said.   
1                  0  @VirginAmerica plus you've added commercials t...   
2                  0  @VirginAmerica I didn't today... Must mean I n...   
3                  0  @VirginAmerica it's really aggressive to blast...   
4                  0  @VirginAmerica and it's a really big bad thing...   
...              ...                                                ...   
14635              0  @AmericanAir thank you we got on a different f...   
14636              0  @AmericanAir leaving over 20 minutes Late Flig...   
14637              0  @AmericanAir Please bring American Airlines to...   
14638              0  @AmericanAir you have my money, you change my ...   
14639              0  @AmericanAir we have 8 ppl so we need 2 know h...   

      tweet_coord              tweet_created tweet_location  \
0             NaN  2015-02-24 11:35:52 -0800            NaN   
1             NaN  2015-02-24 11:15:59 -0800            NaN   
2             NaN  2015-02-24 11:15:48 -0800      Lets Play   
3             NaN  2015-02-24 11:15:36 -0800            NaN   
4             NaN  2015-02-24 11:14:45 -0800            NaN   
...           ...                        ...            ...   
14635   

# Making every column with words lowercase

In [2]:
df['negativereason'] = df['negativereason'].str.lower()
df['airline'] = df['airline'].str.lower()
df['name'] = df['name'].str.lower()
df['text'] = df['text'].str.lower()
df['tweet_location'] = df['tweet_location'].str.lower()
df['user_timezone'] = df['user_timezone'].str.lower()

In [3]:
df

tweet_id airline_sentiment  airline_sentiment_confidence  \
0      570306133677760513           neutral                        1.0000   
1      570301130888122368          positive                        0.3486   
2      570301083672813571           neutral                        0.6837   
3      570301031407624196          negative                        1.0000   
4      570300817074462722          negative                        1.0000   
...                   ...               ...                           ...   
14635  569587686496825344          positive                        0.3487   
14636  569587371693355008          negative                        1.0000   
14637  569587242672398336           neutral                        1.0000   
14638  569587188687634433          negative                        1.0000   
14639  569587140490866689           neutral                        0.6771   

               negativereason  negativereason_confidence         airline  \
0                         NaN                        NaN  virgin america   
1                         NaN                     0.0000  virgin america   
2                         NaN                        NaN  virgin america   
3                  bad flight                     0.7033  virgin america   
4                  can't tell                     1.0000  virgin america   
...                       ...                        ...             ...   
14635                     NaN                     0.0000        american   
14636  customer service issue                     1.0000        american   
14637                     NaN                        NaN        american   
14638  customer service issue                     0.6659        american   
14639                     NaN                     0.0000        american   

      airline_sentiment_gold             name negativereason_gold  \
0                        NaN          cairdin                 NaN   
1                        NaN         jnardino                 NaN   
2                        NaN       yvonnalynn                 NaN   
3                        NaN         jnardino                 NaN   
4                        NaN         jnardino                 NaN   
...                      ...              ...                 ...   
14635                    NaN  kristenreenders                 NaN   
14636                    NaN         itsropes                 NaN   
14637                    NaN         sanyabun                 NaN   
14638                    NaN       srajackson                 NaN   
14639                    NaN        daviddtwu                 NaN   

       retweet_count                                               text  \
0                  0                @virginamerica what @dhepburn said.   
1                  0  @virginamerica plus you've added commercials t...   
2                  0  @virginamerica i didn't today... must mean i n...   
3                  0  @virginamerica it's really aggressive to blast...   
4                  0  @virginamerica and it's a really big bad thing...   
...              ...                                                ...   
14635              0  @americanair thank you we got on a different f...   
14636              0  @americanair leaving over 20 minutes late flig...   
14637              0  @americanair please bring american airlines to...   
14638              0  @americanair you have my money, you change my ...   
14639              0  @americanair we have 8 ppl so we need 2 know h...   

      tweet_coord              tweet_created tweet_location  \
0             NaN  2015-02-24 11:35:52 -0800            NaN   
1             NaN  2015-02-24 11:15:59 -0800            NaN   
2             NaN  2015-02-24 11:15:48 -0800      lets play   
3             NaN  2015-02-24 11:15:36 -0800            NaN   
4             NaN  2015-02-24 11:14:45 -0800            NaN   
...           ...                        ...            ...   
14635   

Printing the dataframe to make sure the values became lowercase

# Removing stop words and punctuation from the tweets

In [4]:
import nltk
from nltk.corpus import stopwords

stop_words = stopwords.words('english')
df['text'] = df['text'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)]))
df['text']

0                           @virginamerica @dhepburn said.
1        @virginamerica plus added commercials experien...
2        @virginamerica today... must mean need take an...
3        @virginamerica really aggressive blast obnoxio...
4                      @virginamerica really big bad thing
                               ...                        
14635     @americanair thank got different flight chicago.
14636    @americanair leaving 20 minutes late flight. w...
14637    @americanair please bring american airlines #b...
14638    @americanair money, change flight, answer phon...
14639    @americanair 8 ppl need 2 know many seats next...
Name: text, Length: 14640, dtype: object

# Finding and filling NA values

In [5]:
df.isna().sum()

tweet_id                            0
airline_sentiment                   0
airline_sentiment_confidence        0
negativereason                   5462
negativereason_confidence        4118
airline                             0
airline_sentiment_gold          14600
name                                0
negativereason_gold             14608
retweet_count                       0
text                                0
tweet_coord                     13621
tweet_created                       0
tweet_location                   4733
user_timezone                    4820
dtype: int64

In [6]:
df = df.fillna(df.mode().iloc[0])

Filled the NA vaules with most common value for each column

In [7]:
df.isna().sum()

tweet_id                        0
airline_sentiment               0
airline_sentiment_confidence    0
negativereason                  0
negativereason_confidence       0
airline                         0
airline_sentiment_gold          0
name                            0
negativereason_gold             0
retweet_count                   0
text                            0
tweet_coord                     0
tweet_created                   0
tweet_location                  0
user_timezone                   0
dtype: int64

Checking to make sure the NA values were filled

# Removing puncuation from the tweets and tokenizing the words

In [8]:
df['text'] = df['text'].str.replace('[^\w\s]','')
df['text']

/var/folders/l3/jbdk0rqd5zn4k08_d788r29h0000gn/T/ipykernel_1596/3644560740.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['text'] = df['text'].str.replace('[^\w\s]','')


0                              virginamerica dhepburn said
1        virginamerica plus added commercials experienc...
2        virginamerica today must mean need take anothe...
3        virginamerica really aggressive blast obnoxiou...
4                       virginamerica really big bad thing
                               ...                        
14635       americanair thank got different flight chicago
14636    americanair leaving 20 minutes late flight war...
14637    americanair please bring american airlines bla...
14638    americanair money change flight answer phones ...
14639    americanair 8 ppl need 2 know many seats next ...
Name: text, Length: 14640, dtype: object

Removing the puncuation from the tweets 

# Using tenssoflow tokenizer to tokenize the text and put it in a new column

In [9]:
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
t = Tokenizer()
t.fit_on_texts(df['text'])
text_tokenized = t.texts_to_sequences(df['text'])

text_tokenized_pad = tf.keras.preprocessing.sequence.pad_sequences(text_tokenized, maxlen=2, truncating='pre')

df2 = pd.concat([df, pd.DataFrame(text_tokenized_pad, columns=['text','tokenized_text'])], axis=1)
df2

tweet_id airline_sentiment  airline_sentiment_confidence  \
0      570306133677760513           neutral                        1.0000   
1      570301130888122368          positive                        0.3486   
2      570301083672813571           neutral                        0.6837   
3      570301031407624196          negative                        1.0000   
4      570300817074462722          negative                        1.0000   
...                   ...               ...                           ...   
14635  569587686496825344          positive                        0.3487   
14636  569587371693355008          negative                        1.0000   
14637  569587242672398336           neutral                        1.0000   
14638  569587188687634433          negative                        1.0000   
14639  569587140490866689           neutral                        0.6771   

               negativereason  negativereason_confidence         airline  \
0      customer service issue                     1.0000  virgin america   
1      customer service issue                     0.0000  virgin america   
2      customer service issue                     1.0000  virgin america   
3                  bad flight                     0.7033  virgin america   
4                  can't tell                     1.0000  virgin america   
...                       ...                        ...             ...   
14635  customer service issue                     0.0000        american   
14636  customer service issue                     1.0000        american   
14637  customer service issue                     1.0000        american   
14638  customer service issue                     0.6659        american   
14639  customer service issue                     0.0000        american   

      airline_sentiment_gold             name     negativereason_gold  \
0                   negative          cairdin  Customer Service Issue   
1                   negative         jnardino  Customer Service Issue   
2                   negative       yvonnalynn  Customer Service Issue   
3                   negative         jnardino  Customer Service Issue   
4                   negative         jnardino  Customer Service Issue   
...                      ...              ...                     ...   
14635               negative  kristenreenders  Customer Service Issue   
14636               negative         itsropes  Customer Service Issue   
14637               negative         sanyabun  Customer Service Issue   
14638               negative       srajackson  Customer Service Issue   
14639               negative        daviddtwu  Customer Service Issue   

       retweet_count                                               text  \
0                  0                        virginamerica dhepburn said   
1                  0  virginamerica plus added commercials experienc...   
2                  0  virginamerica today must mean need take anothe...   
3                  0  virginamerica really aggressive blast obnoxiou...   
4                  0                 virginamerica really big bad thing   
...              ...                                                ...   
14635              0     americanair thank got different flight chicago   
14636              0  americanair leaving 20 minutes late flight war...   
14637              0  americanair please bring american airlines bla...   
14638              0  americanair money change flight answer phones ...   
14639              0  americanair 8 ppl need 2 know many seats next ...   

      tweet_coord              tweet_created  tweet_location  \
0      [0.0, 0.0]  2015-02-24 11:35:52 -0800  washington, dc   
1      [0.0, 0.0]  2015-02-24 11:15:59 -0800  washington, dc   
2      [0.0, 0.0]  2015-02-24 11:15:48 -0800       lets play   
3      [0.0, 0.0]  2015-02-24 11:15:36 -0800  washington, dc   
4      [0.0, 0.0]  2015-02-24 11:14:45 -0800  washington, dc   
...           ...

Created a second dataframe to store the values of the original dataframe and the new tokenized text

In [10]:
df2.drop(['text'], axis =1, inplace = True)
df2

tweet_id airline_sentiment  airline_sentiment_confidence  \
0      570306133677760513           neutral                        1.0000   
1      570301130888122368          positive                        0.3486   
2      570301083672813571           neutral                        0.6837   
3      570301031407624196          negative                        1.0000   
4      570300817074462722          negative                        1.0000   
...                   ...               ...                           ...   
14635  569587686496825344          positive                        0.3487   
14636  569587371693355008          negative                        1.0000   
14637  569587242672398336           neutral                        1.0000   
14638  569587188687634433          negative                        1.0000   
14639  569587140490866689           neutral                        0.6771   

               negativereason  negativereason_confidence         airline  \
0      customer service issue                     1.0000  virgin america   
1      customer service issue                     0.0000  virgin america   
2      customer service issue                     1.0000  virgin america   
3                  bad flight                     0.7033  virgin america   
4                  can't tell                     1.0000  virgin america   
...                       ...                        ...             ...   
14635  customer service issue                     0.0000        american   
14636  customer service issue                     1.0000        american   
14637  customer service issue                     1.0000        american   
14638  customer service issue                     0.6659        american   
14639  customer service issue                     0.0000        american   

      airline_sentiment_gold             name     negativereason_gold  \
0                   negative          cairdin  Customer Service Issue   
1                   negative         jnardino  Customer Service Issue   
2                   negative       yvonnalynn  Customer Service Issue   
3                   negative         jnardino  Customer Service Issue   
4                   negative         jnardino  Customer Service Issue   
...                      ...              ...                     ...   
14635               negative  kristenreenders  Customer Service Issue   
14636               negative         itsropes  Customer Service Issue   
14637               negative         sanyabun  Customer Service Issue   
14638               negative       srajackson  Customer Service Issue   
14639               negative        daviddtwu  Customer Service Issue   

       retweet_count tweet_coord              tweet_created  tweet_location  \
0                  0  [0.0, 0.0]  2015-02-24 11:35:52 -0800  washington, dc   
1                  0  [0.0, 0.0]  2015-02-24 11:15:59 -0800  washington, dc   
2                  0  [0.0, 0.0]  2015-02-24 11:15:48 -0800       lets play   
3                  0  [0.0, 0.0]  2015-02-24 11:15:36 -0800  washington, dc   
4                  0  [0.0, 0.0]  2015-02-24 11:14:45 -0800  washington, dc   
...              ...         ...                        ...             ...   
14635              0  [0.0, 0.0]  2015-02-22 12:01:01 -0800  washington, dc   
14636              0  [0.0, 0.0]  2015-02-22 11:59:46 -0800           texas   
14637              0  [0.0, 0.0]  2015-02-22 11:59:15 -0800   nigeria,lagos   
14638              0  [0.0, 0.0]  2015-02-22 11:59:02 -0800      new jersey   
14639              0  [0.0, 0.0]  2015-02-22 11:58:51 -0800      dallas, tx   

                    user_timezone  tokenized_text  
0      eastern time (us & canada)             137  
1      pacific time (us & canada)            6592  
2      central time (us & canada)             117  
3      pacific time (us & canada)            2666  
4      pacific time (us & canada)             402  
...                           ...        

Droping the text column since it is not tokenized

# Label encoding the airlien_sentiment column

In [11]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df2['airline_sentiment'] = le.fit_transform(df2['airline_sentiment'])
df2

tweet_id  airline_sentiment  airline_sentiment_confidence  \
0      570306133677760513                  1                        1.0000   
1      570301130888122368                  2                        0.3486   
2      570301083672813571                  1                        0.6837   
3      570301031407624196                  0                        1.0000   
4      570300817074462722                  0                        1.0000   
...                   ...                ...                           ...   
14635  569587686496825344                  2                        0.3487   
14636  569587371693355008                  0                        1.0000   
14637  569587242672398336                  1                        1.0000   
14638  569587188687634433                  0                        1.0000   
14639  569587140490866689                  1                        0.6771   

               negativereason  negativereason_confidence         airline  \
0      customer service issue                     1.0000  virgin america   
1      customer service issue                     0.0000  virgin america   
2      customer service issue                     1.0000  virgin america   
3                  bad flight                     0.7033  virgin america   
4                  can't tell                     1.0000  virgin america   
...                       ...                        ...             ...   
14635  customer service issue                     0.0000        american   
14636  customer service issue                     1.0000        american   
14637  customer service issue                     1.0000        american   
14638  customer service issue                     0.6659        american   
14639  customer service issue                     0.0000        american   

      airline_sentiment_gold             name     negativereason_gold  \
0                   negative          cairdin  Customer Service Issue   
1                   negative         jnardino  Customer Service Issue   
2                   negative       yvonnalynn  Customer Service Issue   
3                   negative         jnardino  Customer Service Issue   
4                   negative         jnardino  Customer Service Issue   
...                      ...              ...                     ...   
14635               negative  kristenreenders  Customer Service Issue   
14636               negative         itsropes  Customer Service Issue   
14637               negative         sanyabun  Customer Service Issue   
14638               negative       srajackson  Customer Service Issue   
14639               negative        daviddtwu  Customer Service Issue   

       retweet_count tweet_coord              tweet_created  tweet_location  \
0                  0  [0.0, 0.0]  2015-02-24 11:35:52 -0800  washington, dc   
1                  0  [0.0, 0.0]  2015-02-24 11:15:59 -0800  washington, dc   
2                  0  [0.0, 0.0]  2015-02-24 11:15:48 -0800       lets play   
3                  0  [0.0, 0.0]  2015-02-24 11:15:36 -0800  washington, dc   
4                  0  [0.0, 0.0]  2015-02-24 11:14:45 -0800  washington, dc   
...              ...         ...                        ...             ...   
14635              0  [0.0, 0.0]  2015-02-22 12:01:01 -0800  washington, dc   
14636              0  [0.0, 0.0]  2015-02-22 11:59:46 -0800           texas   
14637              0  [0.0, 0.0]  2015-02-22 11:59:15 -0800   nigeria,lagos   
14638              0  [0.0, 0.0]  2015-02-22 11:59:02 -0800      new jersey   
14639              0  [0.0, 0.0]  2015-02-22 11:58:51 -0800      dallas, tx   

                    user_timezone  tokenized_text  
0      eastern time (us & canada)             137  
1      pacific time (us & canada)            6592  
2      central time (us & canada)             117  
3      pacific time (us & canada)            2666  
4      pacific time (us & canada)             402  
...                          

Assigning 1 to netural, 2 to positive, and 0 to negative

# Splitting the dataset in to train and test

In [12]:
from sklearn.model_selection import train_test_split
train,test = train_test_split(df,test_size=0.2,random_state=42)

# Setting up and scoring attaboost

In [13]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.datasets import make_classification
train, test = make_classification(n_samples=100, n_features=4,n_informative=2, n_redundant=0,random_state=0, shuffle=False)
clf = AdaBoostClassifier(n_estimators=100, random_state=0).fit(train, test)
AdaBoostClassifier(n_estimators=100, random_state=0)
clf.score(train, test)

1.0

The score for attaboost is 1.0

# Setting up and running bagging

In [14]:
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier
from sklearn.datasets import make_classification
train,test = make_classification(n_samples=100, n_features=4,n_informative=2, n_redundant=0,random_state=0, shuffle=False)
clfbagging = BaggingClassifier(base_estimator=SVC(),n_estimators=2, random_state=0).fit(train, test)
clfbagging.score(train, test)

0.89

I made sure to give each of these the same number of n_samples, n_informative, n_redundant, random_state, and Shuffle. Bagging scored .89

# Setting up and running random forest

In [15]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
train, test = make_classification(n_samples=100, n_features=4,n_informative=2, n_redundant=0,random_state=0, shuffle=False)
clfrf = RandomForestClassifier(max_depth=2, random_state=0).fit(train, test)
clfrf.score(train,test)

0.93

Lastly random forest scored .93

Of the 3 Attaboost did the best and had the highest accuracy score and bagging had the lowest. I think attaboost did the best because it generates more than one weak learner and then eventally outputs the one that is most accurate. While random forest and bagging use randomness